### ISSUE TITLE
several crashes with v1.2.1 rc1

### ISSUE THREAD

**Issue by frankystone** [widelands/widelands#6557]
We played a multiplayer game with version 1.2.1-rc1. 

Map: Fellowships
Players: Teayo (release build, windows i guess), hessenfarmer(release build, windows), kaputtnik, host  (debug build, linux)
Win condition: Territorial Lord, 3 hrs
Naval warefare: enabled

After some time the following occurred:
Teayo wasn't able to do create a warship because his ship window didn't provide the "Refit to Warship button". kaputtnik had this button, but as soon he clicked on it the game crashed.
After reloading the save game, we found that the "Configure game" menu lacked the entry "Enable Naval Warefare" below the "Show more options" dropdown.
Starting this save game works fine and Teayo suddenly had the "Refit to warship" button in his ship window and kaputtnik could press this button in his ship window without a crash.

After that from time to time the client hessenfarmer crashed with error "Unreachable code was reached", but after reloading the emergency save game we could continue the game.
After several crashes of this type we decided to end the game.

**Replays**
[2024-10-31T2.zip](https://github.com/user-attachments/files/17594009/2024-10-31T2.zip)

**Savegames**
[wl_autosave_nethost_0.zip](https://github.com/user-attachments/files/17594014/wl_autosave_nethost_0.zip)

**Version:**
 - OS: div
 - Widelands Version: v 1.2.1-rc1
 - Enabled Add-Ons: "Where are my heroes" but i didn't used it (probably because i had no heroes :smile:  )

**Comment by bunnybot** [comment]
Mirrored on Codeberg as [#CB4917](https://codeberg.org/wl/widelands/issues/4917).

**Comment by hessenfarmer** [comment]
the actual error message was this

ERROR:   FATAL EXCEPTION: [D:/a/widelands/widelands/src/logic/map.cc:2191] [D:/a/widelands/widelands/src/logic/map.cc:2191] Unreachable code was reached.
[stdout.txt](https://github.com/user-attachments/files/17594103/stdout.txt)

**Comment by frankystone** [comment]
Ah, tomorrow i try to remember to enable the "write syncstreams" option (if i host again). or should this be enabled by all players?

**Comment by frankystone** [comment]
Hm, i can't find the option "Write Syncstreams" anymore. Was it removed?

**Comment by bunnybot** [comment]
[<img align='left' src='https://codeberg.org/avatars/593dc43e2444c25db9481960e5b693ab' alt='tothxa' height='48' />](https://codeberg.org/tothxa)*[Mirrored from Codeberg](https://codeberg.org/wl/widelands/issues/4917#issuecomment-2418412)*
*On Fri Nov 01 01:20:51 CET 2024, [**Tóth András (tothxa)**](https://codeberg.org/tothxa) wrote:*

---

> Hm, i can't find the option "Write Syncstreams" anymore. Was it removed?

Yes, it became commandline only. Requested in #4733, done in https://github.com/widelands/widelands/pull/6059

**Comment by bunnybot** [comment]
[<img align='left' src='https://codeberg.org/avatars/a4ad2f9a2504f052397d8ba82ed48ba1' alt='Nordfriese' height='48' />](https://codeberg.org/Nordfriese)*[Mirrored from Codeberg](https://codeberg.org/wl/widelands/issues/4917#issuecomment-2418909)*
*On Fri Nov 01 10:31:12 CET 2024, [**Benedikt Straub (Nordfriese)**](https://codeberg.org/Nordfriese) wrote:*

---

Syncstreams are useless for anything but desyncs, and we don't have any of those currently, that's why the option was removed.

---

I ran all four replays with 1.2.1 (Release, Linux) and could not trigger any crashes, they all ran fine. 

The `map.cc:2191` happens when `Map::get_neighbour()` is called with a direction that is none of the six cardinal direction constants (possibly `IDLE` or uninitialized). A backtrace would be helpful.

---

>  After reloading the save game, we found that the "Configure game" menu lacked the entry "Enable Naval Warefare" below the "Show more options" dropdown.

I checked the code quickly and it seems this only happens with savegames, not with new maps, and should be used only for display. The actual value of this flag is read from the savegame during loading. The flag `map_is_seafaring_` is set in `LaunchGame::update_win_conditions()` which only preloads the map file but not a savegame file.

---

I can confirm the bug that naval warfare is not enabled in the client, will investigate.

**Comment by bunnybot** [comment]
Assigned to Nordfriese

**Comment by frankystone** [comment]
> > After reloading the save game, we found that the "Configure game" menu lacked the entry "Enable Naval Warefare" below the "Show more options" dropdown.
> 
> I checked the code quickly and it seems this only happens with savegames, not with new maps, and should be used only for display. The actual value of this flag is read from the savegame during loading. The flag `map_is_seafaring_` is set in `LaunchGame::update_win_conditions()` which only preloads the map file but not a savegame file.

During the game Teayo thought that naval warefare was disabled because he hadn't the button in his ship window. Except that i had this button there was no possibility to check if naval warefare was enabled. I guess that's why hessenfarmer looked into the dropdown after loading the save game. So this issue was encountered.
Not sure, but shouldn't all game settings be visible after loading a save game?

**Comment by bunnybot** [comment]
[<img align='left' src='https://codeberg.org/avatars/a4ad2f9a2504f052397d8ba82ed48ba1' alt='Nordfriese' height='48' />](https://codeberg.org/Nordfriese)*[Mirrored from Codeberg](https://codeberg.org/wl/widelands/issues/4917#issuecomment-2418970)*
*On Fri Nov 01 11:22:54 CET 2024, [**Benedikt Straub (Nordfriese)**](https://codeberg.org/Nordfriese) wrote:*

---

Note that even with this change, loading a savegame does not show the game flags in use. It only shows the relevant checkboxes as disabled with a tooltip "Setting \<X\> is set by the saved game". Changing that would require additionally writing the flags to the savegame preload file, please open a separate feature request for this.

**Comment by hessenfarmer** [comment]
I will be available tonight only from 20:15. 
However I'll try to get a debug build for getting a backtrace probably.

**Comment by hessenfarmer** [comment]
ok we encountered the unreachable code bug now again. 
it can be reproduced only in our 32 bit windows version 

to reproduce 
1. load the savegame 
[2024-11-01T23.48.41.wgf.zip](https://github.com/user-attachments/files/17604745/2024-11-01T23.48.41.wgf.zip)

2. select blue ship near red starting port
3. pres the anchor button.

**Comment by frankystone** [comment]
We could reproduce the crash with this savegame (just rename the extension to .wgf)
[2024-11-01T23.40.26.zip](https://github.com/user-attachments/files/17604725/2024-11-01T23.40.26.zip)

As soon the blue ship "Acua" at position 42,102 starts attacking the red ship "Saledus" the game crashes, or is set to pause. My console output just says: 

~~~~
[04:09:17.501 real] WARNING: [Host]: disconnect_player_controller(0, hessenfarmer)
[04:09:17.513 real] WARNING: [Host]: disconnect_client(0, CLIENT_CRASHED, )
[04:09:21.401 real] INFO: [BufferedConnection] End of file when receiving data from host, closing socket
[04:13:05.585 real] ERROR: ##############################
[04:13:05.585 real] ERROR:   FATAL EXCEPTION: CONNECTION_LOST
[04:13:05.585 real] ERROR: ##############################
[04:13:05.585 real] ERROR:   Please report this problem to help us improve Widelands.
[04:13:05.585 real] ERROR:   You will find related messages in the standard output (stdout.txt on Windows).
[04:13:05.585 real] ERROR:   You are using version 1.2.1~rc1 Debug.
[04:13:05.585 real] ERROR:   Please add this information to your report.
[04:13:05.585 real] ERROR:   If desired, Widelands attempts to create an emergency savegame.
[04:13:05.585 real] ERROR:   It is often – though not always – possible to load it and continue playing.
[04:13:05.585 real] ERROR: ##############################
[04:13:06.699 real] WARNING: [Host]: disconnect_player_controller(0, hessenfarmer)
[04:13:06.718 real] WARNING: [Host]: disconnect_client(0, CONNECTION_LOST, )
[04:13:24.478 real] INFO: [BufferedConnection] End of file when receiving data from host, closing socket
[04:15:57.932 real] INFO: [BufferedConnection] End of file when receiving data from host, closing socket
[04:16:00.523 real] INFO: SoundHandler: Closing 1 time, 44100 Hz, format 32784, 2 channels
[04:16:00.523 real] INFO: SoundHandler: SDL_AUDIODRIVER pulseaudio
~~~~

**Comment by frankystone** [comment]
Maybe it's just because a 32 bit version was running on a 64bit OS?

If hessenfarmer uses a 64 bit version the crash was not reproducible.

**Comment by tothxa** [comment]
I was able to reproduce in `wine`, but I couldn't quite figure out `winedbg`.

**Comment by bunnybot** [comment]
[<img align='left' src='https://codeberg.org/avatars/a4ad2f9a2504f052397d8ba82ed48ba1' alt='Nordfriese' height='48' />](https://codeberg.org/Nordfriese)*[Mirrored from Codeberg](https://codeberg.org/wl/widelands/issues/4917#issuecomment-2420432)*
*On Fri Nov 01 22:42:04 CET 2024, [**Benedikt Straub (Nordfriese)**](https://codeberg.org/Nordfriese) wrote:*

---

LaunchMPG fix cherry-picked in 0190160c6b899d4377d81cec3620401488aff12c

**Comment by hessenfarmer** [comment]
I am afraid the debug symbols are missing.      
will try to compile my own debug build with the symbols and try again

**Comment by tothxa** [comment]
> I am afraid the debug symbols are missing.

Yes, I've checked the workflow files, and the mingw builds strip the exe unconditionally... The msvc build only strips release builds, so I've just extracted the msvc debug build, but it's the same size too... :( Actually not much of a surprise, as the installer sizes are about the same. Oh, wait, msvc puts the debug symbols in a separate .pdb file, doesn't it?

BTW the bug affects the msvc build too.
